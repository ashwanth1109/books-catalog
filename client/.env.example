NEXT_PUBLIC_BOOKS_API_URL="http://localhost:4001/api/books/graphql"
