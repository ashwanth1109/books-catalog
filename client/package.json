{
  "name": "client",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev -p 4000",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "test": "jest"
  },
  "dependencies": {
    "@apollo/client": "3.4.16",
    "@ashwanth1109/books-catalog-common": "1.0.7",
    "antd": "4.16.13",
    "graphql": "15.6.1",
    "next": "11.1.2",
    "nextjs-progressbar": "0.0.11",
    "react": "17.0.2",
    "react-dom": "17.0.2"
  },
  "devDependencies": {
    "@types/react": "17.0.32",
    "autoprefixer": "10.3.7",
    "eslint": "7.32.0",
    "eslint-config-next": "11.1.2",
    "postcss": "8.3.11",
    "tailwindcss": "2.2.17",
    "typescript": "4.4.4"
  }
}
