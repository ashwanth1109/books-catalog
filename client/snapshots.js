module.exports = {
  "__version": "8.7.0",
  "home page e2e": {
    "should assert that books are listed": {
      "1": {
        "data": {
          "books": [
            {
              "_id": "6179a1f524ee6437f46a8eb1",
              "title": "Things Fall Apart",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb2",
              "title": "Fairy tales",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb3",
              "title": "The Divine Comedy",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb4",
              "title": "The Epic Of Gilgamesh",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb5",
              "title": "The Book Of Job",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb6",
              "title": "One Thousand and One Nights",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb7",
              "title": "Njál's Saga",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb8",
              "title": "Pride and Prejudice",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb9",
              "title": "Le Père Goriot",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eba",
              "title": "Molloy, Malone Dies, The Unnamable, the trilogy",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebb",
              "title": "The Decameron",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebc",
              "title": "Ficciones",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebd",
              "title": "Wuthering Heights",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebe",
              "title": "The Stranger",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebf",
              "title": "Poems",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec0",
              "title": "Journey to the End of the Night",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec1",
              "title": "Don Quijote De La Mancha",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec2",
              "title": "The Canterbury Tales",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec3",
              "title": "Stories",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec4",
              "title": "Nostromo",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec5",
              "title": "Great Expectations",
              "__typename": "Book"
            }
          ]
        }
      },
      "2": {
        "data": {
          "book": {
            "_id": "6179a1f524ee6437f46a8eb1",
            "title": "Things Fall Apart",
            "description": "Lorem ipsum is placeholder text commonly used in the graphic, print, and publishing industries for previewing layouts and visual mockups.",
            "year": 1958,
            "__typename": "Book"
          }
        }
      }
    },
    "should assert that books app works as expected": {
      "1": {
        "data": {
          "books": [
            {
              "_id": "6179a1f524ee6437f46a8eb1",
              "title": "Things Fall Apart",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb2",
              "title": "Fairy tales",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb3",
              "title": "The Divine Comedy",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb4",
              "title": "The Epic Of Gilgamesh",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb5",
              "title": "The Book Of Job",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb6",
              "title": "One Thousand and One Nights",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb7",
              "title": "Njál's Saga",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb8",
              "title": "Pride and Prejudice",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eb9",
              "title": "Le Père Goriot",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8eba",
              "title": "Molloy, Malone Dies, The Unnamable, the trilogy",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebb",
              "title": "The Decameron",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebc",
              "title": "Ficciones",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebd",
              "title": "Wuthering Heights",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebe",
              "title": "The Stranger",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ebf",
              "title": "Poems",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec0",
              "title": "Journey to the End of the Night",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec1",
              "title": "Don Quijote De La Mancha",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec2",
              "title": "The Canterbury Tales",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec3",
              "title": "Stories",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec4",
              "title": "Nostromo",
              "__typename": "Book"
            },
            {
              "_id": "6179a1f524ee6437f46a8ec5",
              "title": "Great Expectations",
              "__typename": "Book"
            }
          ]
        }
      },
      "2": {
        "data": {
          "book": {
            "_id": "6179a1f524ee6437f46a8eb1",
            "title": "Things Fall Apart",
            "description": "Lorem ipsum is placeholder text commonly used in the graphic, print, and publishing industries for previewing layouts and visual mockups.",
            "year": 1958,
            "__typename": "Book"
          }
        }
      }
    }
  }
}
