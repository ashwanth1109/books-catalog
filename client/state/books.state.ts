import { BehaviorSubject } from "rxjs";

export interface BookListItem {
  _id: string;
  title: string;
}

export interface Book extends BookListItem {
  description: string;
  year: number;
}

class BooksState {
  public all: BehaviorSubject<BookListItem[]> = new BehaviorSubject<
    BookListItem[]
  >([]);
  public current: BehaviorSubject<Book | null> =
    new BehaviorSubject<Book | null>(null);
}

export default BooksState;
