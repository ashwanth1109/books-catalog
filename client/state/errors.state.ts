import { ReplaySubject } from "rxjs";
import { Logger } from "@ashwanth1109/books-catalog-common";

export interface ClientError {
  message?: string;
  code?: string;
  type?: string;
}

export enum ErrorType {
  GRAPHQL_CLIENT_ERROR = "GRAPHQL_CLIENT_ERROR",
}

export enum ErrorCode {
  NETWORK_ERROR = "NETWORK_ERROR",
  CLIENT_ERROR = "CLIENT_ERROR",
}

class ErrorsState {
  // We prefer replay subject in case we need to explore history of errors
  public error: ReplaySubject<ClientError> = new ReplaySubject<ClientError>(5);

  constructor() {
    this.error.subscribe((error) => {
      Logger.error("Ran into error: ", error?.message);
      // TODO: Throw a toast error here with "error?.message"
      // TODO: send platform errors to some event bus for observability
      // This way, we will be notified of errors even before users report it
      // With all the necessary debugging information
      // This also lets us know if we have any buggy releases recently.
    });
  }
}

export default ErrorsState;
