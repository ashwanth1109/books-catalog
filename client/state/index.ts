import { createContext } from "react";

import BooksState from "./books.state";
import ErrorsState from "./errors.state";

// Keeping state as a class instance =>
// this allows for maintaining more than one instance of state in the application (if necessary)
class State {
  public booksState: BooksState;
  public errorsState: ErrorsState;

  constructor() {
    this.booksState = new BooksState();
    this.errorsState = new ErrorsState();
  }
}

export default State;

export const initialState = new State();
export const AppContext = createContext<State>(initialState);
