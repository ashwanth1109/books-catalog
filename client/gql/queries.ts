import { gql } from "@apollo/client";

export const GET_BOOK_BY_ID = gql`
  query GetBookById($bookId: ObjectId!) {
    book(bookId: $bookId) {
      _id
      title
      description
      year
    }
  }
`;

export const GET_ALL_BOOKS = gql`
  query GetAllBooks {
    books {
      _id
      title
    }
  }
`;
