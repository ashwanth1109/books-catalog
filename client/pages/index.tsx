import { Card } from "antd";
import { useContext, useEffect } from "react";

import type { NextPage } from "next";

import CustomBreadcrumb from "../components/custom-breadcrumb.component";
import BookList from "../components/book-list.component";
import useGqlQueryHook from "../hooks/use-gql-query.hook";
import { AppContext } from "../state";
import { GET_ALL_BOOKS } from "../gql/queries";

/**
 * FIXME: Initial hydration of data should happen on server side in "getInitialProps" to take advantage of SSR
 * However, the challenge is requests made from server do not get forwarded to ingress from within the cluster.
 * We need to setup networking for enabling inter-cluster communication
 * Options: Containers in the same pod? Sidecar pods?
 */

const Home: NextPage = () => {
  const { loading, data } = useGqlQueryHook(GET_ALL_BOOKS);
  const { booksState } = useContext(AppContext);

  useEffect(() => {
    booksState.all.next(data?.books || []);
  }, [booksState, data]);

  return (
    <>
      <CustomBreadcrumb />
      <Card
        title="List of books in your catalog:"
        className="p-8"
        loading={loading}
      >
        <BookList />
      </Card>
    </>
  );
};

export default Home;
