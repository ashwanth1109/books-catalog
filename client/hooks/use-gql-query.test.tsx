import { act, renderHook } from "@testing-library/react-hooks";
import { ReactNode, useContext } from "react";
import { QueryHookOptions } from "@apollo/client";
import { MockedProvider } from "@apollo/client/testing";
import { MockedResponse } from "@apollo/client/utilities/testing/mocking/mockLink";

import useGqlQueryHook from "hooks/use-gql-query.hook";
import { GET_ALL_BOOKS, GET_BOOK_BY_ID } from "gql/queries";
import { MapAny } from "../types/util.types";
import { expect } from "@jest/globals";
import { AppContext } from "../state";
import { ReplaySubject } from "rxjs";
import { ClientError, ErrorCode, ErrorType } from "../state/errors.state";
import { GraphQLError } from "graphql";

const BOOK_ID = "1234";

const wrapWithMockProvider = (gqlMocks: ReadonlyArray<MockedResponse>) => {
  return ({ children }: { children: ReactNode }) => (
    <MockedProvider mocks={gqlMocks} addTypename={false}>
      {children}
    </MockedProvider>
  );
};

describe("use-gql-query hook", () => {
  describe("loading state", () => {
    const wrapper = wrapWithMockProvider([
      {
        request: { query: GET_BOOK_BY_ID, variables: { bookId: BOOK_ID } },
      },
    ]);

    it("should assert that loading state is returned when gql is loading", async () => {
      await act(async () => {
        const { result } = renderHook(
          () =>
            useGqlQueryHook(GET_BOOK_BY_ID, {
              bookId: BOOK_ID,
            } as QueryHookOptions<MapAny, MapAny>),
          {
            wrapper,
          }
        );

        expect(result.current.loading).toBe(true);
        expect(result.current.data).toBeUndefined();
      });
    });
  });

  describe("network error", () => {
    const wrapper = wrapWithMockProvider([
      {
        request: { query: GET_BOOK_BY_ID, variables: { bookId: BOOK_ID } },
        error: new Error("Test error"),
      },
    ]);

    it("should assert that network errors are sent to error state", async () => {
      const errorSubject = new ReplaySubject<ClientError>();

      await act(async () => {
        renderHook(
          () => {
            const { errorsState } = useContext(AppContext);
            errorsState.error = errorSubject;
            useGqlQueryHook(GET_BOOK_BY_ID, {
              bookId: BOOK_ID,
            } as QueryHookOptions<MapAny, MapAny>);
          },
          {
            wrapper,
          }
        );

        await new Promise((resolve) => setTimeout(resolve, 0)); // wait for response
        await new Promise((resolve) => {
          errorSubject.subscribe((error) => {
            expect(error.type).toBe(ErrorType.GRAPHQL_CLIENT_ERROR);
            expect(error.code).toBe(ErrorCode.NETWORK_ERROR);
            expect(error.message).toMatchSnapshot();
            resolve(void 0);
          });
        });
      });
    });
  });

  describe("graphql errors", () => {
    const testCode = "testCode";
    const testError = "testError";
    const wrapper = wrapWithMockProvider([
      {
        request: { query: GET_ALL_BOOKS },
        result: {
          errors: [
            new GraphQLError(testError, null, null, null, null, null, {
              code: testCode,
            }),
          ],
        },
      },
    ]);

    it("should assert that graphql errors are sent to error state", async () => {
      const errorSubject = new ReplaySubject<ClientError>();

      await act(async () => {
        renderHook(
          () => {
            const { errorsState } = useContext(AppContext);
            errorsState.error = errorSubject;
            useGqlQueryHook(GET_ALL_BOOKS);
          },
          { wrapper }
        );

        await new Promise((resolve) => setTimeout(resolve, 0)); // wait for response
        await new Promise((resolve) => {
          errorSubject.subscribe((error) => {
            console.log(error);
            expect(error.type).toBe(ErrorType.GRAPHQL_CLIENT_ERROR);
            expect(error.message).toBe(testError);
            expect(error.code).toBe(testCode);
            resolve(void 0);
          });
        });
      });
    });
  });
});
