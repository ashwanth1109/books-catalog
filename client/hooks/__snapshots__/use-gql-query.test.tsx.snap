// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`use-gql-query hook network error should assert that network errors are sent to error state 1`] = `
"No more mocked responses for the query: query GetBookById($bookId: ObjectId!) {
  book(bookId: $bookId) {
    _id
    title
    description
    year
  }
}

Expected variables: {}

Failed to match 1 mock for this query, which had the following variables:
  {\\"bookId\\":\\"1234\\"}
"
`;
